File: packages/vscode-extension/src/providers/base/operations.ts
/**
 * Base Operation Types
 * 
 * Operations represent user actions that modify the schema state.
 * All operations follow an event-sourcing pattern and are immutable.
 */

/**
 * Base operation structure
 */
export interface Operation {
  /** Unique identifier for this operation */
  id: string;
  
  /** ISO timestamp when operation was created */
  ts: string;
  
  /** Provider ID (e.g., 'unity', 'hive', 'postgres') */
  provider: string;
  
  /** Operation type with provider prefix (e.g., 'unity.add_catalog') */
  op: string;
  
  /** ID of the target object this operation applies to */
  target: string;
  
  /** Operation-specific payload data */
  payload: Record<string, any>;
}

/**
 * Operation category for grouping related operations
 */
export enum OperationCategory {
  Catalog = 'catalog',
  Schema = 'schema',
  Table = 'table',
  Column = 'column',
  Constraint = 'constraint',
  Security = 'security',
  Metadata = 'metadata',
}

/**
 * Operation metadata for UI and documentation
 */
export interface OperationMetadata {
  /** Operation type (e.g., 'unity.add_catalog') */
  type: string;
  
  /** Human-readable display name */
  displayName: string;
  
  /** Description of what this operation does */
  description: string;
  
  /** Category this operation belongs to */
  category: OperationCategory;
  
  /** Required payload fields */
  requiredFields: string[];
  
  /** Optional payload fields */
  optionalFields: string[];
  
  /** Whether this operation is destructive (drops/deletes) */
  isDestructive: boolean;
}

/**
 * Helper to create a new operation with defaults
 */
export function createOperation(
  provider: string,
  opType: string,
  target: string,
  payload: Record<string, any>,
  id?: string
): Operation {
  return {
    id: id || `op_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
    ts: new Date().toISOString(),
    provider,
    op: `${provider}.${opType}`,
    target,
    payload,
  };
}

File: packages/python-sdk/src/schematic/providers/base/operations.py
"""
Base Operation Types

Operations represent user actions that modify the schema state.
"""

import random
import string
import time
from enum import StrEnum
from typing import Any

from pydantic import BaseModel


class OperationCategory(StrEnum):
    """Operation category for grouping"""

    CATALOG = "catalog"
    SCHEMA = "schema"
    TABLE = "table"
    COLUMN = "column"
    CONSTRAINT = "constraint"
    SECURITY = "security"
    METADATA = "metadata"


class Operation(BaseModel):
    """Base operation structure"""

    id: str  # Unique identifier
    ts: str  # ISO timestamp
    provider: str  # Provider ID (e.g., 'unity', 'hive')
    op: str  # Operation type with provider prefix
    target: str  # ID of target object
    payload: dict[str, Any]  # Operation-specific data


class OperationMetadata(BaseModel):
    """Operation metadata for UI and documentation"""

    type: str  # Operation type
    display_name: str  # Human-readable name
    description: str  # Description
    category: OperationCategory  # Category
    required_fields: list[str]  # Required payload fields
    optional_fields: list[str]  # Optional payload fields
    is_destructive: bool  # Whether operation is destructive


def create_operation(
    provider: str,
    op_type: str,
    target: str,
    payload: dict[str, Any],
    op_id: str | None = None,
) -> Operation:
    """
    Helper to create a new operation with defaults

    Args:
        provider: Provider ID
        op_type: Operation type (without provider prefix)
        target: Target object ID
        payload: Operation payload
        op_id: Optional operation ID (generated if not provided)

    Returns:
        Operation instance
    """
    if op_id is None:
        # Generate a unique ID
        timestamp = int(time.time() * 1000)
        random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=9))
        op_id = f"op_{timestamp}_{random_suffix}"

    return Operation(
        id=op_id,
        ts=time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime()),
        provider=provider,
        op=f"{provider}.{op_type}",
        target=target,
        payload=payload,
    )
